axiom-eth/src/rlp/mod.rs

```diff
@@ -625,4 +807,268 @@ impl<'range, F: ScalarField> RlpChip<'range, F> {
+        self.range.check_less_than_safe(
+            ctx,
+            list_payload_len,
+            (max_rlp_array_len - max_len_len) as u64,
+        );
+
+        // this is automatically <= max_rlp_array_len
+        // right now assuming it's a list, so it's ok
+        // if short: 1 + len
+        // if long: 1 + len_len + len
+        let rlp_len = self
+            .gate()
+            .sum(ctx, [Constant(F::one()), Existing(len_len), Existing(list_payload_len)]);
+
+        let mut item_witness = Vec::with_capacity(max_item_lens.len());
+
+        let mut prefix_idx = self.gate().add(ctx, Constant(F::one()), Existing(len_len));
+        let mut running_max_len = max_len_len + 1; 
+
+        for &max_item_len in max_item_lens {
+            // Every item is either:
+            // phantom (an empty filler, if the RLP list has fewer items than max_item_lens.len()),
+            // literal (a single byte literal between 0x00 and 0x7f),
+            // short (a byte array or list of length <= 55), or
+            // long (a byte array or list of length > 55)
+
+            // if phantom: prefix == 0
+            // otherwise: first byte of item, to be parsed
+            let mut prefix = self.gate().select_from_idx(
+                ctx,
+                rlp_array.iter().copied().take(running_max_len + 1),
+                prefix_idx,
+            );
+            let prefix_parsed = self.parse_rlp_prefix(ctx, prefix); // constrained correct
+
+            let mut len_len = prefix_parsed.len_len;
+            let max_item_len_len = max_rlp_len_len(max_item_len);
+            self.range.check_less_than_safe(ctx, len_len, (max_item_len_len + 1) as u64);
+            
+            let len_start_id = *prefix_parsed.is_not_literal.value() + prefix_idx.value();
+            // len_cells is empty if phantom, literal, or short
+            let len_cells = witness_subarray(
+                ctx,
+                &rlp_array,
+                &len_start_id,
+                len_len.value(),
+                max_item_len_len,
+            );   // ***** unconstrained
+
+            // 0 if phantom, literal or short
+            // len if long
+            let len_byte_val = evaluate_byte_array(ctx, self.gate(), &len_cells, len_len);
+                // constrained
+
+            // len if short or long
+            // 1 if phantom or literal
+            let mut item_len = self.gate().select(
+                ctx,
+                len_byte_val,
+                prefix_parsed.next_len,
+                prefix_parsed.is_big,
+            );
+
+
+            // prefix_len is 0 if phantom or literal
+            // 1 otherwise
+            let mut prefix_len = prefix_parsed.is_not_literal;
+
+            if is_variable_len{
+                // If `prefix_idx >= rlp_len`, that means we are done
+                // Question: can we change this to an equality check?
```

Comment:
I don't think so? since you need it to be true for every non-phantom item

Reply:
Sorry -- I meant an inequality check!
I think if it's phantom, then you'll have equality `prefix_idx == rlp_len`, so we just have to check if they are equal or not (which is cheaper than a range check)...